linkerd/dns/src/lib.rs

```diff
@@ -3,45 +3,60 @@
 mod refine;
 
 pub use self::refine::{MakeRefine, Refine};
+use async_trait::async_trait;
 pub use linkerd2_dns_name::{InvalidName, Name, Suffix};
 use std::future::Future;
 use std::pin::Pin;
 use std::{fmt, net};
 use tokio::sync::{mpsc, oneshot};
+use tokio::time::Instant;
 use tracing::{info_span, trace, Span};
 use tracing_futures::Instrument;
 pub use trust_dns_resolver::config::ResolverOpts;
 pub use trust_dns_resolver::error::{ResolveError, ResolveErrorKind};
-use trust_dns_resolver::lookup_ip::LookupIp;
+pub use trust_dns_resolver::lookup_ip::LookupIp;
 use trust_dns_resolver::{config::ResolverConfig, system_conf, AsyncResolver};
 
+#[derive(Clone, Debug)]
+pub struct ResolveResponse {
+    pub ips: Vec<net::IpAddr>,
+    pub valid_until: Instant,
+}
+
+struct ResolveRequest {
+    name: Name,
+    result_tx: oneshot::Sender<Result<LookupIp, ResolveError>>,
+    span: tracing::Span,
+}
+
+#[async_trait]
+pub trait Resolver: Clone {
```

Comment:
What is this trait used for? Where are we generic over the DNS impl?

Reply:
In the tests

https://github.com/linkerd/linkerd2-proxy/blob/33d4ec8aa2a3e081ea286f24573119fb7fb4fcaf/linkerd/app/core/src/control.rs#L294

Reply:
Hm, it feels like this shouldn't be tied to the DNS behavior bout to resolution behavior in general...

Reply:
But we are testing the logic that turns a Dns resolution request/response API into a resolution stream. So not sure how we can abstract that or avoid the trait. Maybe I am missing something

Reply:
This set of tests is here to test the logic in this function: https://github.com/linkerd/linkerd2-proxy/pull/594/files/33d4ec8aa2a3e081ea286f24573119fb7fb4fcaf#diff-fe9e02d6254c4455e6c23df861eb03d1R222

Reply:
Maybe another way to approach this... Can we do something similar without having to introduce a new trait? It looks like the tests only mock a single async function. What if, instead of the dns-specific trait, we model this use case as a `Service` that is either implemented by or adaptable from the dns resolver type?

Reply:
Got it! Thanks for the tip!